UVC: move application decision to the application

josuah opened the pull request:
Dependency:

- https://github.com/zephyrproject-rtos/zephyr/pull/92320

The USB class was doing arbitrary "best guess" choices about what pixel formats or resolutions to include (all of them, and min/max only when format are ranges).

- In the class, add a `uvc_add_format()` to allow the application to push the exact format it wants
- In the class, remove all format-related video API calls away from the UVC stack, to let the application handle this
- In the sample, filter out formats that are not supported by the host whenever standard formats are also supported
- In the sample, filter out formats that are too large for the memory

What is added in the sample is a matter of application choices, and trapping these choices in the UVC class prevented the samples to work without manual configuration on diverse platform.

MacOS support still not confirmed.
Linux, Windows, Android tested functional with the hardware I have.

josuah committed:
usb: device_next: uvc: remove application decisions from the UVC class

The UVC class was deciding itself which formats were sent to the host.
Remove this logic out of the UVC class and introduce uvc_add_format() to
give the application the freedom of which format to list.

Signed-off-by: Josuah Demangeon <me@josuah.net>

josuah committed:
samples: usb: uvc: add filtering of the format

The UVC class now lets the application select the format list sent to the
host. Leverage this in the sample to filter out any format that is not
expected to work (buffer too big, rarely supported formats).

Signed-off-by: Josuah Demangeon <me@josuah.net>

josuah commented:
This is not ready for review, will be properly split in individual commits, and "un-drafted" when ready...

josuah commented:
Force-push:
- Rebase on top of the MacOSX-related bugfix

Windows, Linux, MacOSX, Android are now expected to all work.
Tested with nRF52840-DK:

```
west build --board=nrf52840dk/nrf52840 --snippet=video-sw-generator samples/subsys/usb/uvc/
```

josuah committed:
doc: release: 4.3: add UVC changes

Add USB UVC device's new uvc_add_format() function to the release note,
and document the semantic changes of UVC APIs in the migration guide.

Signed-off-by: Josuah Demangeon <me@josuah.net>

tmon-nordic reviewed (dismissed):
Needs rebase.

tmon-nordic commented on samples/subsys/usb/uvc/src/main.c:
```diff
@@ -17,15 +17,66 @@
+const static struct device *const uvc_dev = DEVICE_DT_GET(DT_NODELABEL(uvc));
+const static struct device *const video_dev = DEVICE_DT_GET(DT_CHOSEN(zephyr_camera));
+
+/* Format capabilities of video_dev, usd everywhere through the sampel */
```
sampel -> sample

tmon-nordic commented on samples/subsys/usb/uvc/src/main.c:
```diff
@@ -32,7 +32,28 @@ static size_t app_get_min_buf_size(const struct video_format *const fmt)
+/* Check whether the video device supports one of the wisespread image sensor formats */
+static bool app_has_standard_formats(void)
+{
+	for (int i = 0;; i++) {
```
Is this loop guaranteed to not come out of bounds?

avolmat-st reviewed (dismissed):
LGTM, thanks a lot !!

josuah committed:
usb: device_next: uvc: propagate descriptor error to the application

When running out of descriptor, return an error instead of ignoring it.
The application need to make sure to adjust the Kconfig macros to have
enough descriptors for all formats to add.

Signed-off-by: Josuah Demangeon <me@josuah.net>

josuah committed:
usb: device_next: uvc: use fmt->size instead of computing it every time

Make use of the recently merged fmt->size to know the maximum size of the
frame to be allocated, which works for both compressed and uncompressed
video.

Signed-off-by: Josuah Demangeon <me@josuah.net>

ngphibang reviewed (commented):
Just some nits, otherwise LGTM.

josuah reviewed (commented):
Force-push:
- Rebased on `main`
- Applied @ngphibang  suggestions

josuah reviewed (commented):
.

ngphibang reviewed (dismissed):
LGTM. Thanks

avolmat-st commented:
Hi @jfischer-no, could you have a look again ? Thanks.

jfischer-no commented on subsys/usb/device_next/class/usbd_uvc.c:
```diff
@@ -1674,22 +1664,13 @@ static int uvc_init(struct usbd_class_data *const c_data)
-			if (ret != 0) {
-				return ret;
-			}
+			uvc_add_vs_format_desc(dev, &format_desc, cap);
```
I am not sure it makes sense. If due to wrong configuration the UVC function will be non functional, then it should fail during initialization. Otherwise I think it is okay but I would keep the return values above, and then log them and not fail here.

jfischer-no replied:
But I am still curious, will the UVC be functional if CONFIG_USBD_VIDEO_MAX_FORMATS or CONFIG_USBD_VIDEO_MAX_FRMIVAL are set too low?

josuah replied:
It would present fewer formats to the host, and the host would still be able to select some.

It could eventually lead to one invalid format descriptor with no frame/resolution descriptor for it, or it could lead to a list of descriptors that lack a few supported format but still valid as per USB standard.

I could as well return the error directly if this makes more sense to ensure "descriptors guaranteed correct, or nothing at all".

Guidance appreciated!

avolmat-st reviewed (dismissed):
LGTM. Thanks.

jfischer-no commented on subsys/usb/device_next/class/usbd_uvc.c:
```diff
@@ -1500,23 +1500,22 @@ static void uvc_set_vs_bitrate_range(struct uvc_frame_discrete_descriptor *const
 
 	uvc_set_vs_bitrate_range(desc, video_frmival_nsec(frmival), fmt);
-
-	return 0;
```
This commit does not build. Missed return will cause build warning.

jfischer-no commented on subsys/usb/device_next/class/usbd_uvc.c:
```diff
@@ -1624,6 +1642,7 @@ static int uvc_init(struct usbd_class_data *const c_data)
 	struct video_caps caps;
 	uint32_t prev_pixfmt = 0;
 	uint32_t mask = 0;
+	int first_err = ret;
```
Compile error, that does not work in C. `int first_err = 0;` ?

josuah replied:
Adjusted in latest git push, sorry.

jfischer-no commented on subsys/usb/device_next/class/usbd_uvc.c:
```diff
@@ -1671,34 +1690,50 @@ static int uvc_init(struct usbd_class_data *const c_data)
-				uvc_assign_desc(dev, &cfg->desc->if1_color, true, true);
+
+				ret = uvc_assign_desc(dev, &cfg->desc->if1_color, true, true);
+				if (ret != 0) && first_err == 0) {
```
compile error
```c

				ret = uvc_assign_desc(dev, &cfg->desc->if1_color, true, true);
				if (ret != 0) && first_err == 0) {
--------------------------------------------^
```

jfischer-no commented on subsys/usb/device_next/class/usbd_uvc.c:
```diff
@@ -1712,7 +1747,7 @@ static int uvc_init(struct usbd_class_data *const c_data)
 	atomic_set_bit(&data->state, UVC_STATE_INITIALIZED);
 
-	return 0;
+	return first_err;
```
I do not entirely get the idea here. Is the user really care about what the first error was?

josuah replied:
The idea was to allow a few functions to fail due to being short on memory for the descriptors. In this case some errors would be ignored, with the error code stored for later. We then have the choice: return the first or last error code encountered.

But it is also possible to fail early, which would prevent the enumeration to happen when running out of configured memory. This would be a lot less confusing to debug and avoid such oddities.

josuah commented:
nRF52840 with `-DCONFIG_USBD_VIDEO_MAX_FORMATS=1` (completely invalid, should not work):

```
*** Booting Zephyr OS build v4.2.0-6382-g042a6e7b3bb7 ***
[00:00:00.250,183] <wrn> usbd_uvc: Not enough format descriptors to add descriptors for 'YUYV' and 64x64
[00:00:00.250,213] <wrn> uvc_sample: Skipping format 1920x1080
[00:00:00.250,366] <err> usbd_uvc: Invalid format ID (1) and/or frame ID (1)
[00:00:00.250,366] <err> usbd_uvc: init: failed to query the default probe
[00:00:00.250,366] <err> usbd_init: Failed to initialize class instance
[00:00:00.250,396] <err> usbd_init: Failed to init FS configuration 1
[00:00:00.250,457] <err> usbd_sample_config: Failed to initialize device support
```

nRF52840 with `-DCONFIG_USBD_VIDEO_MAX_FORMATS=2` (just enough for video-sw-generator):

```
*** Booting Zephyr OS build v4.2.0-6382-g042a6e7b3bb7 ***
[00:00:00.250,213] <wrn> uvc_sample: Skipping format 1920x1080
[00:00:00.250,457] <inf> uvc_sample: Waiting the host to select the video format
```

Here on FRDM-MCXN947 with `-DCONFIG_USBD_VIDEO_MAX_FORMATS=2`:

![mpv-shot0006](https://github.com/user-attachments/assets/e7e4d7c4-ecee-49c6-8b6f-e9d677532141)

avolmat-st reviewed (approved):
LGTM. Tested ok with PR#94562 on top of it via the STM32N6.

josuah commented on subsys/usb/device_next/class/usbd_uvc.c:
```diff
@@ -1552,7 +1549,7 @@ static int uvc_add_vs_frame_desc(const struct device *dev,
 	desc->wWidth = sys_cpu_to_le16(w);
 	desc->wHeight = sys_cpu_to_le16(h);
-	desc->dwMaxVideoFrameBufferSize = sys_cpu_to_le32(max_size);
+	desc->dwMaxVideoFrameBufferSize = sys_cpu_to_le32(fmt.size);
```
I must have verified the final version of the file, but this `fmt` struct is local in this commit. Good catch for bissectability!

josuah replied:
It looks like this got merged. Since this got raised past-review and is in git history by now. One commit will have a runtime (not compile time) error unfortunately :disappointed:
